src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets

```diff
@@ -102,6 +102,53 @@ Copyright (c) .NET Foundation. All rights reserved.
     <RuntimeFrameworkVersion Condition="'$(RuntimeFrameworkVersion)' == ''">$(_TargetFrameworkVersionWithoutV)</RuntimeFrameworkVersion>
   </PropertyGroup>
 
+  <ItemGroup Condition="'$(SetLinkMetadataAutomatically)' != 'false'">
+    <Compile Update="@(Compile)">
+      <Link Condition="'%(Link)' == '' And !$([System.IO.Path]::GetFullPath('%(FullPath)').StartsWith('$([System.IO.Path]::GetFullPath('$(MSBuildProjectDirectory)'))\'))">*NeedsLink*</Link>
+
+      <Link Condition="'%(Link)' == '*NeedsLink*' And '%(RecursiveDir)' != '' And '%(LinkBase)' == ''">%(RecursiveDir)%(Filename)%(Extension)</Link>
+      <Link Condition="'%(Link)' == '*NeedsLink*' And '%(RecursiveDir)' != '' And '%(LinkBase)' != ''">%(LinkBase)\%(RecursiveDir)%(Filename)%(Extension)</Link>
+      <Link Condition="'%(Link)' == '*NeedsLink*' And '%(RecursiveDir)' == '' And '%(LinkBase)' == ''">%(Filename)%(Extension)</Link>
+      <Link Condition="'%(Link)' == '*NeedsLink*' And '%(RecursiveDir)' == '' And '%(LinkBase)' != ''">%(LinkBase)\%(Filename)%(Extension)</Link>
```

Comment:
Can you just use %(RecursiveDir) without checking if it's empty. Seems it would evaluate to the right thing in both cases.